File: arcgis_map_sdk_android/android/src/main/kotlin/dev/fluttercommunity/arcgis_map_sdk_android/util/GraphicsParser.kt
package dev.fluttercommunity.arcgis_map_sdk_android.util

import android.graphics.drawable.BitmapDrawable
import com.arcgismaps.Color
import com.arcgismaps.geometry.Point
import com.arcgismaps.geometry.Polygon
import com.arcgismaps.geometry.Polyline
import com.arcgismaps.geometry.SpatialReference
import com.arcgismaps.mapping.symbology.PictureMarkerSymbol
import com.arcgismaps.mapping.symbology.SimpleFillSymbol
import com.arcgismaps.mapping.symbology.SimpleLineSymbol
import com.arcgismaps.mapping.symbology.SimpleLineSymbolStyle
import com.arcgismaps.mapping.symbology.SimpleMarkerSymbol
import com.arcgismaps.mapping.symbology.Symbol
import com.arcgismaps.mapping.view.Graphic
import dev.fluttercommunity.arcgis_map_sdk_android.model.LatLng
import dev.fluttercommunity.arcgis_map_sdk_android.model.symbol.PictureMarkerSymbolPayload
import dev.fluttercommunity.arcgis_map_sdk_android.model.symbol.SimpleFillSymbolPayload
import dev.fluttercommunity.arcgis_map_sdk_android.model.symbol.SimpleLineSymbolPayload
import dev.fluttercommunity.arcgis_map_sdk_android.model.symbol.SimpleMarkerSymbolPayload
import dev.fluttercommunity.arcgis_map_sdk_android.model.toAGSPoint
import dev.fluttercommunity.arcgis_map_sdk_android.parseToClass
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import java.io.InputStream


class GraphicsParser(private val binding: FlutterPluginBinding) {

    fun parse(map: Map<String, Any>): List<Graphic> {
        val graphics = when (val type = map["type"] as String) {
            "point" -> parsePoint(map)
            "polygon" -> parsePolygon(map)
            "polyline" -> parsePolyline(map)
            else -> throw Exception("No type for $type")
        }

        val attributes = map["attributes"] as? Map<String, String>

        if (attributes != null) {
            return graphics.map {
                attributes.forEach { (key, value) ->
                    it.attributes[key] = value
                }
                it
            }
        }

        return graphics
    }

    private fun parsePoint(map: Map<String, Any>): List<Graphic> {
        val point = (map["point"] as Map<String, Any>).parseToClass<LatLng>()
        val symbolMap = map["symbol"] as Map<String, Any>

        val pointGraphic = Graphic().apply {
            geometry = point.toAGSPoint()
            symbol = parseSymbol(symbolMap)
        }

        return listOf(pointGraphic)
    }

    private fun parsePolyline(map: Map<String, Any>): List<Graphic> {
        val points = parseToClass<List<List<List<Double>>>>(map["paths"]!!)
        val symbolMap = map["symbol"] as Map<String, Any>

        return points.map { subPoints ->
            Graphic().apply {
                geometry = Polyline(subPoints.map { coordinateArray ->
                    val x = coordinateArray.elementAtOrNull(0)
                    val y = coordinateArray.elementAtOrNull(1)
                    val z = coordinateArray.elementAtOrNull(2)
                    if (x == null || y == null) {
                        throw Exception("Coordinate array needs at least 2 doubles. Got $coordinateArray")
                    }

                    if (z != null) Point(x, y, z, SpatialReference.wgs84())
                    else Point(x, y, SpatialReference.wgs84())
                })
                symbol = parseSymbol(symbolMap)
            }
        }
    }

    private fun parsePolygon(map: Map<String, Any>): List<Graphic> {
        val rings = parseToClass<List<List<List<Double>>>>(map["rings"]!!)
        val symbolMap = map["symbol"] as Map<String, Any>

        return rings.map { ring ->
            Graphic().apply {
                geometry = Polygon(ring.map { LatLng(it[0], it[1]).toAGSPoint() })
                symbol = parseSymbol(symbolMap)
            }
        }
    }

    fun parseSymbol(symbolMap: Map<String, Any>): Symbol {
        val symbol = when (val type = symbolMap["type"]) {
            "simple-marker" -> parseSimpleMarkerSymbol(symbolMap)
            "picture-marker" -> parsePictureMarkerSymbol(symbolMap)
            "simple-fill" -> parseSimpleFillSymbol(symbolMap)
            "simple-line" -> parseSimpleLineSymbol(symbolMap)
            else -> throw Exception("No type for $type")
        }

        return symbol
    }

    private fun parseSimpleMarkerSymbol(map: Map<String, Any>): Symbol {
        val payload = map.parseToClass<SimpleMarkerSymbolPayload>()

        return SimpleMarkerSymbol().apply {
            color = Color(payload.color.toHexInt())
            size = payload.size.toFloat()
            outline = SimpleLineSymbol().apply {
                style = SimpleLineSymbolStyle.Solid
                color = Color(payload.outlineColor.toHexInt())
                width = payload.outlineWidth.toFloat()
            }
        }
    }

    private fun parsePictureMarkerSymbol(map: Map<String, Any>): Symbol {
        val payload = map.parseToClass<PictureMarkerSymbolPayload>()

        // return local asset in case its a local path
        if (!payload.assetUri.isWebUrl()) {
            return PictureMarkerSymbol.createWithImage(getBitmapFromAssetPath(payload.assetUri)!!)
                .apply {
                    width = payload.width.toFloat()
                    height = payload.height.toFloat()
                    offsetX = payload.xOffset.toFloat()
                    offsetY = payload.yOffset.toFloat()
                }
        }

        return PictureMarkerSymbol(payload.assetUri).apply {
            width = payload.width.toFloat()
            height = payload.height.toFloat()
            offsetX = payload.xOffset.toFloat()
            offsetY = payload.yOffset.toFloat()
        }
    }

    private fun getBitmapFromAssetPath(asset: String): BitmapDrawable? {
        val assetPath: String = binding.flutterAssets.getAssetFilePathBySubpath(asset)

        var inputStream: InputStream? = null
        val drawable: BitmapDrawable?
        try {
            inputStream = binding.applicationContext.assets.open(assetPath)
            drawable = BitmapDrawable.createFromStream(inputStream, assetPath) as BitmapDrawable
        } finally {
            inputStream?.close()
        }
        return drawable
    }

    private fun parseSimpleFillSymbol(map: Map<String, Any>): Symbol {
        val payload = map.parseToClass<SimpleFillSymbolPayload>()

        return SimpleFillSymbol().apply {
            color = Color(payload.fillColor.toHexInt())
            outline = SimpleLineSymbol().apply {
                style = SimpleLineSymbolStyle.Solid
                color = Color(payload.outlineColor.toHexInt())
                width = payload.outlineWidth.toFloat()
            }
        }
    }

    private fun parseSimpleLineSymbol(map: Map<String, Any>): Symbol {
        val payload = map.parseToClass<SimpleLineSymbolPayload>()

        return SimpleLineSymbol().apply {
            if (payload.color != null) color = Color(payload.color.toHexInt())
            payload.marker?.let {
                markerStyle = it.style
                markerPlacement = it.placement
            }

            style = payload.style
            width = payload.width.toFloat()
        }

    }
}

fun String.isWebUrl(): Boolean = this.startsWith("https://") || this.startsWith("http://")

File: arcgis_map_sdk_ios/ios/Classes/GraphicsParser.swift
//
//  GraphicsParser.swift
//  arcgis_map_sdk_ios
//
//  Created by Julian Bissekkou on 09.12.22.
//

import Foundation
import ArcGIS

class GraphicsParser {
    let registrar: FlutterPluginRegistrar

    init(registrar: FlutterPluginRegistrar) {
            self.registrar = registrar
        }


    func parse(dictionary: Dictionary<String, Any>) throws -> [AGSGraphic] {
        let type = dictionary["type"] as! String

        let newGraphics: [AGSGraphic]
        switch (type) {
        case "point":
            newGraphics = try! parsePoint(dictionary)
        case "polygon":
            newGraphics = try! parsePolygon(dictionary)
        case "polyline":
            newGraphics = try! parsePolyline(dictionary)
        default:
            throw ParseException(message: "Unknown graphic type: \(type)")
        }

        let attributes = dictionary["attributes"] as? Dictionary<String, Any>
        if let attributes = attributes {
            newGraphics.forEach {
                $0.attributes.addEntries(from: attributes)
            }
        }
        return newGraphics
    }

    private func parsePoint(_ dictionary: [String: Any]) throws -> [AGSGraphic] {
        let point: LatLng = try! JsonUtil.objectOfJson(dictionary["point"] as! Dictionary<String, Any>)

        let graphic = AGSGraphic()
        graphic.geometry = point.toAGSPoint()
        graphic.symbol = try! parseSymbol(dictionary["symbol"] as! Dictionary<String, Any>)

        return [graphic]
    }

    private func parsePolyline(_ dictionary: [String: Any]) throws -> [AGSGraphic] {
        let payload: PathPayload = try! JsonUtil.objectOfJson(dictionary)

        return try payload.paths.map { coordinates in
            let points = try coordinates.map { array in
                if(array.count < 2) {
                    throw ParseException(message: "Size of coordinates need to be at least 2. Got \(array)")
                }
                if(array.count > 2) {
                    return AGSPoint(x: array[0], y: array[1], z: array[2], spatialReference: .wgs84())
                }
                return AGSPoint(x: array[0], y: array[1], spatialReference: .wgs84())
            }

            let graphic = AGSGraphic()
            graphic.geometry = AGSPolyline(points: points)
            graphic.symbol = try! parseSymbol(dictionary["symbol"] as! Dictionary<String, Any>)

            return graphic
        }
    }

    private func parsePolygon(_ dictionary: [String: Any]) throws -> [AGSGraphic] {
        let payload: PolygonPayload = try! JsonUtil.objectOfJson(dictionary)

        return payload.rings.map { ring in
            let graphic = AGSGraphic()
            let points = ring.map { coordinate in
                AGSPoint(x: coordinate[0], y: coordinate[1], spatialReference: .wgs84())
            }
            graphic.geometry = AGSPolygon(points: points)
            graphic.symbol = try! parseSymbol(dictionary["symbol"] as! Dictionary<String, Any>)

            return graphic
        }
    }

    // region symbol parsing

    private func parseSymbol(_ dictionary: [String: Any]) throws -> AGSSymbol {
        let type = dictionary["type"] as! String;
        switch (type) {
        case "simple-marker":
            return parseSimpleMarkerSymbol(dictionary)
        case "picture-marker":
            return parsePictureMarkerSymbol(dictionary)
        case "simple-fill":
            return parseSimpleFillMarkerSymbol(dictionary)
        case "simple-line":
            return parseSimpleLineSymbol(dictionary)
        default:
            throw ParseException(message: "Unknown symbol type: \(type)")
        }
    }

    private func parseSimpleMarkerSymbol(_ dictionary: [String: Any]) -> AGSSymbol {
        let payload: SimpleMarkerSymbolPayload = try! JsonUtil.objectOfJson(dictionary)

        let symbol = AGSSimpleMarkerSymbol()
        symbol.color = payload.color.toUIColor()
        symbol.size = payload.size
        symbol.outline = AGSSimpleLineSymbol(
                style: .solid,
                color: payload.outlineColor.toUIColor(),
                width: payload.outlineWidth
        )
        return symbol
    }

    private func parseSimpleFillMarkerSymbol(_ dictionary: [String: Any]) -> AGSSymbol {
        let payload: SimpleFillSymbolPayload = try! JsonUtil.objectOfJson(dictionary)

        let symbol = AGSSimpleFillSymbol()
        symbol.color = payload.fillColor.toUIColor()

        let outline = AGSSimpleLineSymbol()
        outline.width = payload.outlineWidth
        outline.color = payload.outlineColor.toUIColor()
        symbol.outline = outline

        return symbol
    }

    private func parsePictureMarkerSymbol(_ dictionary: [String: Any]) -> AGSSymbol {
        let payload: PictureMarkerSymbolPayload = try! JsonUtil.objectOfJson(dictionary)

        if(!payload.assetUri.isWebUrl()) {
            let uiImage = getFlutterUiImage(payload.assetUri)
            let symbol = AGSPictureMarkerSymbol(image: uiImage!)
            symbol.width = payload.width
            symbol.height = payload.height
            symbol.offsetX = payload.xOffset
            symbol.offsetY = payload.yOffset
            return symbol
        }

        let symbol = AGSPictureMarkerSymbol(url: URL(string: payload.assetUri)!)
        symbol.width = payload.width
        symbol.height = payload.height
        symbol.offsetX = payload.xOffset
        symbol.offsetY = payload.yOffset

        return symbol
    }

    private func getFlutterUiImage(_ fileName: String) -> UIImage? {
            let key = registrar.lookupKey(forAsset: fileName)
        let path = Bundle.main.path(forResource: key, ofType: nil)
        return UIImage(named: path!)
    }

    private func parseSimpleLineSymbol(_ dictionary: [String: Any]) -> AGSSymbol {
        let payload: SimpleLineSymbolPayload = try! JsonUtil.objectOfJson(dictionary)
        let symbol = AGSSimpleLineSymbol()
        
        if let color = payload.color {
            symbol.color = color.toUIColor()
        }
        if let marker = payload.marker {
            symbol.markerStyle = marker.style.toAGSStyle()
            symbol.markerPlacement = marker.placement.toAGSStyle()
        }

        symbol.style = payload.style.toAGSStyle()
        symbol.width = payload.width

        return symbol
    }

    // endregion
}

private struct PathPayload: Codable {
    let paths: [[[Double]]]
}

private struct PolygonPayload: Codable {
    let rings: [[[Double]]]
}

extension String {
    func isWebUrl()-> Bool {
        return starts(with: "https://") || starts(with: "http://")
    }
}
